src/checkers/inference/BaseInferrableChecker.java

```diff
@@ -0,0 +1,79 @@
+package checkers.inference;
+
+import org.checkerframework.common.basetype.BaseAnnotatedTypeFactory;
+import org.checkerframework.framework.flow.CFAnalysis;
+import org.checkerframework.framework.flow.CFStore;
+import org.checkerframework.framework.flow.CFTransfer;
+import org.checkerframework.framework.flow.CFValue;
+import org.checkerframework.framework.type.GenericAnnotatedTypeFactory;
+import org.checkerframework.javacutil.Pair;
+
+import java.util.List;
+
+import javax.lang.model.element.VariableElement;
+
+import checkers.inference.dataflow.InferenceAnalysis;
+import checkers.inference.dataflow.InferenceTransfer;
+
+import com.sun.source.tree.Tree;
+import com.sun.source.util.Trees;
+
+/**
+ * Default implementation of InferrableChecker.
+ *
+ */
+public abstract class BaseInferrableChecker extends InferenceChecker implements InferrableChecker {
+
+    @Override
+    public void initChecker() {
+        //In between these brackets, is code copied directly from SourceChecker
+        //except for the last line assigning the visitor
+        {
+            Trees trees = Trees.instance(processingEnv);
+            assert( trees != null ); /*nninvariant*/
+            this.trees = trees;
+
+            this.messager = processingEnv.getMessager();
+            this.messages = getMessages();
+
+            this.visitor = createVisitor(null, createRealTypeFactory(), false);
+        }
+    }
+
+    @Override
+    public InferenceVisitor<?, ?> createVisitor(InferenceChecker ichecker, BaseAnnotatedTypeFactory factory, boolean infer) {
+        return new InferenceVisitor<>(this, ichecker, factory, infer);
+    }
+
+    @Override
+    public BaseAnnotatedTypeFactory createRealTypeFactory() {
```

Comment:
The "real" type factory is the one that contains the type hierarchy for which we are inferring annotations (e.g. OsTrustedTypeFactory rather InferenceTypeFactory).

The InferenceTypeFactory is used in Inference mode whereas the real typefactory is used in type checking mode (and by the InferenceTypeFactory during inference mode).